### ISSUE TITLE
Continued server issues (R package)

### ISSUE THREAD

**Issue by j-sayers** [MotusDev/Motus-TO-DO#875]
Over the last few days there have been numerous messages from people unable to download data with the R package. I've been having the same issues. There have been a few periods where data downloads have been possible, but generally attempts to download just haven't been working. Below is the typical pattern. 

> Checking for new data in project 339
> Project   339:  got   335 batch records
> batchID  1346036 (#     1 of    335): got     78 runs starting at               0
> The server did not respond within 120s. Trying again...
> Error: The server is not responding, please try again later.
> Called from: srvQuery(API = motus_vars$API_HITS_FOR_TAG_PROJECT, params = list(projectID = projectID, 
>     batchID = batchID, hitID = hitID), verbose = verbose)

> Checking for new data in project 213
> Project   213:  got  3280 batch records
> batchID   948780 (#     1 of   3280): got      2 runs starting at               0
> The server did not respond within 120s. Trying again...
> Error: The server is not responding, please try again later.
> Called from: srvQuery(API = motus_vars$API_HITS_FOR_TAG_PROJECT, params = list(projectID = projectID, 
>     batchID = batchID, hitID = hitID), verbose = verbose)
>

**Comment by jsayers-bsc** [comment]
Is this related to #874? This did both seem to start roughly the same time, though the Motus site has been working for me throughout (if slower than usual) but I have been unable to download data through the R package (except for a small window during which I unfortunately passed along the premature good news to other people who were having issues).

**Comment by jsayers-bsc** [comment]
In addition to the errors above referring to API = motus_vars$API_HITS_FOR_TAG_PROJECT, the following error occasionally is seen

> Error: Motus API request failed [500]
> Internal Server Error
> Called from: srvQuery(API = motus_vars$API_DEVICE_ID_FOR_RECEIVER, params = list(serno = serno), 
>     verbose = verbose)

Typically though, downloading a receiver's data is more likely to work whereas downloading a project's data nearly always fails.

**Comment by denislepage** [comment]
I have added a NOLOCK option in the 2 hit views.  This allows the users to read the data while inserts or edits may be happening elsewhere in the table, rather than have to wait in line. That may or may not explain the timeouts, but that may help reduce them. I ran the query locally for that batch, and it responds quite fast now, so the indexing seems fine.

@cjardine-bsc  we probably want to expand the use of nolock in most of the R view. I haven't touched sandbox either.

**Comment by cjardine-bsc** [comment]
Hey D-Boss, 

We discussed NOLOCK in the R views in the past. We actually had added into all the views but then removed it. You wanted to try using NOLOCK in Java instead of the views themselves (see below). 

@pmorrill did you ever investigate this? I just want to follow up on this approach since I think Denis has some good justification below for doing this in Java over SQL. Alternatively I'm happy to add it to the views. The risks are low as these views are only used for output. We just need to be sure we keep it that way 👍 

From D-Boss circa January 2018:

Before we start making the change in all the R views though, I am wondering if it wouldn't be easier to make this an option in the Java code (NOLOCK is transitional and applies to all tables within the view you call). The benefit is that we could more easily toggle this as an option without having to rewrite all the views.

Paul, I have modified the method MotusAPIContext.getSGViewName(String name) to ctx.getSGViewName(String name, String alias) and included the NOLOCK option there. Have a look and feel free to change the specifics if you want. It's in beta only for now.

I'd still go for using NOLOCK from Java. Paul, have you had a chance to have a look at my code? CJ, I think that using that option in views is calling for potential trouble. The R views maybe less so as they are rather specific to R and should be output only, but let's try the Java approach first, which is also easier to implement (and already done, in theory). Because NOLOCK is transitional, you should only need to call it once, and it'll be easy to toggle off if we don't like it.

**Comment by pmorrill** [comment]
No follow up yet. I only see issues if assigned to me (or mentioned as you did, C).

Not sure about this in Denis' statement: `It's in beta only for now.` Did you mean sandbox? That's where we usually put new code.

I will track it down when I have a moment.

**Comment by cjardine-bsc** [comment]
This was from an email (back in the dark ages!). I can forward it to you if you want the context.

**Comment by jsayers-bsc** [comment]
I've been unable to download data from the R package for much of the day today due to repeated server timeouts.

I get there are multiple potential causes for that but as there appears to be a potential improvement not yet implemented I'm posted here.

As a general comment, the server timeout error in R is a pretty regular issue; seems like every few days there is a period when I encounter it. Usually I move on to something else and by the time I get back I can download again, so most of these issues never get reported here, despite their frequency. It's also the case that most collaborators aren't trying to download very often so it doesn't affect others as much.
